### PR TITLE
Add special character escaping when facts loaded into test variant

### PR DESCRIPTION
**Pull request by hack-beard**

### CODE REVIEW

**Comment by hack-beard** [comment]
Added function to facts to return a string with special characters escaped for sh, psh, cmd. This modified the associated service function to call this when facts are loaded into test variants to prevent special characters from being passed to command line.

**Comment by privateducky** [comment]
good change set. can you give an example where an ability (or fact) fails without this in place? i may add a unit test to go alongside this.

**Comment by hack-beard** [comment]
Adjusted character set a bit and added to the re.sub() call to escape '\' and everything seems to be functioning as intended on this end now. Does depend on the fix for sandcat to properly preference the executors. Running with the CMD executor in a PSH environment will not work.

**Comment by hack-beard** [comment]
Also, for clarification, tests made using the 'stage sensitive files' ability

File: app/objects/secondclass/c_fact.py [hunk 1]
```diff
@@ -1,5 +1,22 @@
 from app.utility.base_object import BaseObject
 
+escape_ref = {
```
Comment (wbooth) [inline app/objects/secondclass/c_fact.py]: Do you think there is a better place or structure for this information?

**Comment by privateducky** [comment]
i feel good about this. can we add a unit test on the escaped function? put it in a new test class in the tests/ directory. 

what do you mean by "Does depend on the fix for sandcat to properly preference the executors. Running with the CMD executor in a PSH environment will not work."?

**Comment by hack-beard** [comment]
@privateducky the escape characters are set based on the executor. there was a "bug" where agents run in powershell were defaulting to the "cmd" executor. In this case the special character escape function was using the 'cmd' escape character '^' instead of the powershell one '`', even though it was running in powershell. There was a fix up for this at the time these changes were being made, but it had not been merged in yet to my knowledge

**Comment by hack-beard** [comment]
Travis did not approve of the test file that makes sense, that is why the 'sh' test case looks strange

**Pull request merged**